### DISCUSSION TITLE
Better tree-shaking and bundling patterns

### DISCUSSION THREAD

**Discussion by jsudelko** [melt-ui/melt-ui#635] [General]
The current builder pattern and event system used by Melt UI is a breath of fresh air compared to the current state of components in the Svelte ecosystem. That said, I posit that there are quite a bit of byte savings to be had. In particular, as a user, I'd like to only pay for features that I use. For example, if I'm strictly using controlled functionality, the uncontrolled logic (e.g. close action in Popover) is dead weight.

1. Prefer tree-shakable sub-builders
Instead of:
```javascript
import { createDropdownMenu } from '@melt-ui/svelte';

const {
	elements: { trigger, menu, item, separator, arrow },
	// I might not need these
	builders: { createSubmenu, createMenuRadioGroup, createCheckboxItem },
	states: { open },
} = createDropdownMenu({
	forceVisible: true,
	loop: true,
});
```

Do:
```javascript
import { createDropdownMenu, createDropdownMenu, createSubmenu, createCheckboxItem } from '@melt-ui/svelte';
// or import { createDropdownMenu, createSubmenu, createCheckboxItem  } from '@melt-ui/svelte/checkbox';

const {
	elements: { trigger, menu, item, separator, arrow },
	states: { open },
} = createDropdownMenu({
	forceVisible: true,
	loop: true,
});
```

2. Preferring imports over configuration where possible:

Instead of:
```html
<script>
  import { createComponent, melt } from '@melt-ui/svelte';
 
  const {
    elements: { foo, bar },
    values: { baz },
  } = createComponent({
    activateFeature: true,
  });
</script>
```

Do:
```html
<script>
import { createComponent, melt, componentFeature } from '@melt-ui/svelte';
// or import { createComponent, feature } from '@melt-ui/svelte/component';
 
  const {
    elements: { foo, bar },
    values: { baz },
  } = createComponent({
    plugin: [componentFeature] // either this
  });
</script>

<!-- Or this -->
<div use:componentFeature></div>
```

3. Prefer stores unions to derived stores:
Instead of `isChecked: boolean` and `isIndeterminate: boolean`, prefer `checked: true | false | 'indeterminate'` and let the user derive the values if they need to. As a user, I likely won't use some derived stores provided by Melt...I shouldn't have to pay for them. In Svelte 5, deriving values becomes even easier reducing the need for built-in derived helpers.

...you can probably think of a lot more others ways to save bytes if you're familiar with the details of implementation. I open this as a discussion with the hope that more specific issues can spring from this.

The best analogy I can give to motivate this change in mindset is that Melt UI is currently like [`zod`](https://github.com/colinhacks/zod) while it should strive to be more like [`valibot`](https://github.com/fabian-hiller/valibot#comparison). It's not monstrously large but it could be A LOT smaller if composable.

**Comment by TGlide** [comment]
Hi, it seems I've never answered this in full!

It's very much likely that in Melt v2, we'll be separating builders, utilities, etc into their own packages. Which means less disk space to worry about. 

Regarding a plugin system though, at first glance I'm not too much in favour of it. It's odd from an API/DX perspective, it's not obvious which plugins may fit there, and the internal code for constructing the builders around this logic would also be quite confusing I imagine. Especially when sometimes we have defaults, e.g. we default to using Floating UI, where with a plugin system, it'd be harder to choose a default.

I think its similar to a component. You wouldn't, I presume, use plugins to delimit core functionality and configuration. You'd just expose a prop.

**Reply by buhodev** [reply]
What do you mean by 'separating builders, utilities, etc into their own packages'? Having to install @melt-ui/dropdown, @melt-ui/popover, etc? If that is the case, you might have to take into account that it's very easy to have two or more incompatible dependency versions installed at the same time

**Reply by TGlide** [reply]
You can specify peerDeps. But builders are independent, this shouldn't be worrying.

**Reply by jsudelko** [reply]
I made the above suggestions knowing full well that it's difficult to accomplish with your current architecture. It would likely need quite a bit of refactoring & redesigning. Probably some breaking pre-1.0 changes. Because of this, I understand where you're coming from and can sympathize if you decided to keep the status quo

That said, my experience with Melt is that most of the time you should not be using it directly inside other components because it gets messy _fast_. I typically create a styled component on top of Melt and then reuse it elsewhere. Since components are created once with incremental modifications in the future, a slight increase in external complexity is probably not a huge deal. It's a trade-off between one-time developer convenience and significant long-running user benefits from decreased bundle size + improved code splitting. Case in point, I use Valibot for most things now even though you could argue it has a worse DX than Zod.

**Reply by jsudelko** [reply]
On the matter of splitting off components, if I need to install 15 components independently in before I can get started or need to migrate to a major version, that is a significantly worse DX. Especially because I imagine the vast majority of people will use more than half a dozen Melt components. The unpacked version of Melt is 1mb on NPM and will probably never grow much larger...maybe 5mb if you implemented a tone more components...so disk space is not an issue. FWIW, see [this](https://twitter.com/diegohaz/status/1719037919933972729). Please don't make me debug peer dependency errors. 🙈

**Reply by buhodev** [reply]
> FWIW, see [this](https://twitter.com/diegohaz/status/1719037919933972729)

This is exactly the tweet I was looking for but couldn't find :)

**Reply by TGlide** [reply]
It's less about the cost of refactoring (as we've undergone big refactors (see #234) when we thought it was worth it), and more that I just don't think the pros outweigh the cons.

With zod and valibot, the logic is basically the same. You just get the methods from somewhere else, but in the end, `z` from zod is just a catch-all for all methods. With Melt? The changes would be quite a lot. It's not only more verbose, but hard to reason about. Are plugins unique or shared? Which ones fit in which builder? Why? Can I duplicate plugins? Should all configuration go into plugins? If not, why not?

And it's not only about external API, a plugin system, especially for building components, would be quite complex. This makes it harder to maintain, harder to get new contributions, and these are things I value a lot. All for potential marginal bundle size improvements. Sure, some builders could potentially be quite reduced, such as the ones that use Floating UI. But most people will (and should) use it, so in the end, I don't think there would be a lot of benefit to it.

---

That being said, I don't think it's a worthless idea! I know I may sound harsh, but it's just because I don't think it makes sense for Melt, at least currently. It is an interesting approach, but it requires a lot of experimentation. Could even be a separate project. Me personally, I don't want to invest in this road for Melt, for the time being. I could be proven wrong in the future though, I'm no Messiah! I appreciate the discussion and the wishes to further improve Melt though.

**Reply by TGlide** [reply]
> On the matter of splitting off components, if I need to install 15 components independently in before I can get started or need to migrate to a major version, that is a significantly worse DX. 

I agree that a manual install one by one would be tedious. There are several ways around this though. Docs that provide a single copy-paste command based on which components you want, a CLI, etc. This is not a hard problem to solve.

> disk space is not an issue

Until you have tons of projects that use it. Sure Melt alone won't deplete your free space, but these things add up.

Regardless though, I agree disk space is not a big problem! The main reason I'd like to separate these builders into separate packages is so that you're not forced to update all other builders, when you just want to update one. This will be especially helpful when changing major versions, because of breaking changes.

It's also easier to determine what the impact to bundle size is per individual builder.

That being said, I don't feel that strongly about it either. I understand the counter-argument, specially when breaking changes probably won't happen frequently post-v1. This is not set in stone.

**Reply by TGlide** [reply]
P.S. we can do what React Aria does. Ship both individual packages, or a catch-all single package.

**Reply by jsudelko** [reply]
https://valibot.dev/blog/valibot-v0.31.0-is-finally-available/ ❤️

**Reply by jsudelko** [reply]
https://twitter.com/devongovett/status/1813451655465382015

**Reply by TGlide** [reply]
Since you pinged, an update on my end. We'll not be shipping multiple packages anymore. Indeed it gives much more complexity for little gain.

I am also looking into better tree shaking patterns, not necessarily plugin based in most cases though. Most builders configs are really minimal, but I'm exploring :)

**Reply by TGlide** [reply]
Btw, why did you link the second tweet?
